migrations/BatchProcessor.php

```diff
@@ -0,0 +1,257 @@
+<?php
+/**
+ * Class to handle process large amount of data with batch.
+ *
+ * @package Tutor
+ * @author Themeum <support@themeum.com>
+ * @link https://themeum.com
+ * @since 3.8.0
+ */
+
+namespace Tutor\Migrations;
+
+/**
+ * Class BatchProcessor
+ */
+abstract class BatchProcessor {
+	/**
+	 * Batch size.
+	 *
+	 * @var int
+	 */
+	protected $batch_size;
+
+	/**
+	 * Schedule interval
+	 *
+	 * @var int
+	 */
+	protected $schedule_interval;
+
+	/**
+	 * Progress option name to keep track of each process status.
+	 *
+	 * @var string
+	 */
+	protected $progress_option;
+
+	/**
+	 * Manage child class as singleton behavior.
+	 *
+	 * @var object
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Action name to invoke wp-cron.
+	 *
+	 * @var string
+	 */
+	protected $action;
+
+	/**
+	 * Name of the process.
+	 *
+	 * @var string
+	 */
+	protected $name;
+
+	/**
+	 * Constructor.
+	 *
+	 * @throws \Exception If action is not set.
+	 */
+	protected function __construct() {
+		if ( empty( $this->action ) ) {
+			throw new \Exception( 'Action property must be defined in the subclass.' );
+		}
+
+		$this->progress_option = "tutor_batch_processor_{$this->action}";
+		add_action( $this->action, array( $this, 'process_batch' ) );
+	}
+
+	/**
+	 * Get child class instance.
+	 *
+	 * @return object
+	 */
+	public static function instance() {
+		if ( null === static::$instance ) {
+			static::$instance = new static();
+		}
+		return static::$instance;
+	}
+
+	/**
+	 * Get items to process.
+	 * Must be implemented in child class.
+	 *
+	 * @param int $offset offset.
+	 * @param int $limit limit.
+	 *
+	 * @return array
+	 */
+	abstract protected function get_items( $offset, $limit);
```

Comment:
All the **Abstract** methods should have argument & return strict type so that the concrete classes implement it properly and the `process_batch` method can work as expected.